examples/switching_linear_dynamic_model.py

```diff
@@ -0,0 +1,231 @@
+        #
+        ssm_init = pyro.sample("ssm_init", dist.Normal(0,1).expand_by([n_lat,1]).independent(2))
+
+        lengths = lengths[batch]
+        z = 0
+        x = ssm_init
+        for t in range(lengths.max()):
+            with poutine.mask(mask=(t < lengths).unsqueeze(-1)):
+                z = pyro.sample("z_{}".format(t), dist.Categorical(hmm_dyn[z]),
+                                infer={"enumerate": "parallel"})
+                x = pyro.sample("x_{}".format(t), dist.Normal(torch.matmul(ssm_dyn[z], x) + ssm_bias[z], torch.sqrt(1./ssm_noise[z])).independent(2))
+                obs_lat = torch.matmul(obs_weight[z], x) + obs_bias[z]
+                pyro.sample("y_{}".format(t), dist.Normal(obs_lat, torch.sqrt(1./obs_noise[z])).independent(2),
+                            obs=sequences[batch, t].unsqueeze(-1))
+
+
+                # print('===={}===='.format(t))
+                # print('obs_lat', obs_lat.shape)
+                # print('obs_noise[z]', obs_noise[z].shape)
+                # print('sequences[batch, t].unsqueeze(-1).shape', sequences[batch, t].unsqueeze(-1).shape)
+                #
+                # print('obs_bias[z]', obs_bias[z].shape)
+                # print('obs_weight[z]', obs_weight[z].shape)
+                #
+                # print('z_shape:', z.shape)
+                # print('hmm_dyn[z].shape', hmm_dyn[z].shape)
+                # print('cat_event_shape', dist.Categorical(hmm_dyn[z]).event_shape)
+                # print('cat_batch_shape', dist.Categorical(hmm_dyn[z]).batch_shape)
+                #
+                # print('x_shape:', x.shape)
+                # print('x_batch_shape:', dist.Normal(torch.matmul(ssm_dyn[z], x) + ssm_bias[z], torch.sqrt(1. / ssm_noise[z])).independent(2).batch_shape)
+                #
+                # print('y_batch_shape', dist.Normal(obs_lat, torch.sqrt(1./obs_noise[z])).independent(2).batch_shape)
+                # print('y_event.shape', dist.Normal(obs_lat, torch.sqrt(1./obs_noise[z])).independent(2).event_shape)
+                # print('obs_final.shape', sequences[batch, t].unsqueeze(-1).shape)
+
+
+
+def main(args):
+    if args.cuda:
+        torch.set_default_tensor_type('torch.cuda.FloatTensor')
+
+    logging.info('Generating data')
+    logging.info(args)
+
+    ########################################################################
+    # just generate a random dataset for debug purpose
+    ########################################################################
+    N = 3  # number of sequences
+    T = 10 # number of timestamps
+    D = 2
+    sequences = torch.tensor(np.random.randn(N,T,D), dtype=torch.float32)
+    logging.info('-' * 40)
+    model = model_slds
+    lengths = torch.ones(sequences.shape[0], dtype=torch.long)*T
+
+
+    ###################
+    # guide for MAP
+    ###################
+    myguide_map = AutoDelta(poutine.block(model, hide_fn=lambda m: m['name'].startswith('z_')))
```

Comment:
I suspect it won't work well to MAP estimate the `x` state, but it may work well to MAP estimate the global parameters and estimate `x` as `AutoDiagonalNormal` or `AutoLowRankMultivariateNormal`, e.g.:
```py
guide_map = AutoGuideList(model)
guide_map.add(AutoDelta(poutine.block(model, hide_fn=lambda m: re.match('[zx]_', m['name'])))
guide_map.add(AutoDiagonalNormal(poutine.block(model, expose_fn=lambda m: re.match('x_', m['name'])))
```